plugins/glustershd/actor.go

```diff
@@ -19,69 +21,148 @@ const (
 
 type shdActor struct{}
 
-func (actor *shdActor) Do(v *volume.Volinfo, key string, value string, logger log.FieldLogger) error {
-	if key != selfHealKey {
+func (actor *shdActor) Do(v *volume.Volinfo, key string, value string, volOp xlator.VolumeOpType, logger log.FieldLogger) error {
+	voltype := v.Type.String()
+
+	if voltype != "Replicate" && voltype != "Disperse" {
 		return nil
 	}
+
 	glustershDaemon, err := newGlustershd()
 	if err != nil {
 		return err
 	}
-
-	switch value {
-	case "on":
-		if isHealEnabled(v) {
-			err = daemon.Start(glustershDaemon, true, logger)
-			if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+	switch volOp {
+	case xlator.VolumeStart:
+		for k, v := range v.Options {
+			_, _, key, err := options.SplitKey(k)
+			if err != nil {
 				return err
 			}
+			if key == selfHealKey && v == "off" {
+				return nil
+			}
 		}
-	case "off":
-		if !isHealEnabled(v) {
-			isVolRunning, err := volume.AreReplicateVolumesRunning()
-			if err != nil {
+		err := daemon.Start(glustershDaemon, true, logger)
+		if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+			return err
+		}
+
+	case xlator.VolumeStop:
+		isVolRunning, err := volume.AreReplicateVolumesRunning(v.ID)
+		if err != nil {
+			return err
+		}
+
+		if !isVolRunning {
+			err := daemon.Stop(glustershDaemon, true, logger)
+			if err != nil && err != gderrors.ErrPidFileNotFound {
 				return err
 			}
-			if !isVolRunning {
-				return daemon.Stop(glustershDaemon, true, logger)
+		}
+
+	case xlator.VolumeSet:
+		fallthrough
+	case xlator.VolumeReset:
+		if key != selfHealKey {
+			return nil
+		}
+
+		switch value {
+		case "on":
+			if isHealEnabled(v) {
+				err = daemon.Start(glustershDaemon, true, logger)
+				if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+					return err
+				}
+			}
+		case "off":
+			if !isHealEnabled(v) {
+				isVolRunning, err := volume.AreReplicateVolumesRunning(v.ID)
+				if err != nil {
+					return err
+				}
+				if !isVolRunning {
+					err := daemon.Stop(glustershDaemon, true, logger)
+					if !os.IsNotExist(err) {
+						return err
+					}
+				}
 			}
+
 		}
 	}
 	return nil
 }
 
-func (actor *shdActor) Undo(v *volume.Volinfo, key string, value string, logger log.FieldLogger) error {
-	if key != selfHealKey {
+func (actor *shdActor) Undo(v *volume.Volinfo, key string, value string, volOp xlator.VolumeOpType, logger log.FieldLogger) error {
+	voltype := v.Type.String()
+
+	if voltype != "Replicate" && voltype != "Disperse" {
 		return nil
 	}
 
 	glustershDaemon, err := newGlustershd()
 	if err != nil {
 		return err
 	}
-
-	switch value {
-	case "off":
-		if !isHealEnabled(v) {
-			if v.State != volume.VolStarted {
-				return errors.New("volume should be in started state")
-			}
-			err = daemon.Start(glustershDaemon, true, logger)
-			if err != nil && err != gderrors.ErrProcessAlreadyRunning {
+	switch volOp {
+	case xlator.VolumeStart:
+		for k, v := range v.Options {
```

Comment:
Same comment here

Reply:
v.Options may contain the (self-heal)key if the user gave it during volume create. or else start the daemon based on the default option.

Reply:
sure i will update this also